### PR TITLE
2.4 LP:1723184 Hook Error When Unit Not Leader

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

If the uniter queues a _leader-elected_ hook, but application leadership changes before it is run, the unit enters an error state.

This patch modifies the `runHook.Prepare` method to recheck for leadership when running a _leader-elected_ hook. If the unit is no longer the leader, `ErrSkipExecute` is returned.

It also skips the call to `Merge` leadership settings and logs a warning when the unit is not the leader, instead of returning an error.

## QA steps

I tried for some time without success to replicate the issue using the charm in the LP bug, _ubuntu-repository-cache_.

Modified test verifies new behaviour, and error returns for the other failure conditions remain.

To check for regression:
1. Bootstrap to LXD.
2. `juju deploy ubuntu-repository-cache -n 5`.
3. Stop the leader's container with `lxc stop {container-id}`.
4. Check that leadership changes without error and restart the container.
5. Repeat steps 3-4 several times.

## Documentation changes

None. 

## Bug reference

https://bugs.launchpad.net/juju-core/+bug/1723184

### CODE REVIEW

**Comment by howbazaar** [comment]
I am wondering if it would not be better to do the check for leadership before we start to run the hook. That way if the unit is no longer the leader when it wants to run the queued hook, it won't even start. It is possible that the charm is trying to do some things on the assumption that it is the leader at the workload level.

File: worker/uniter/operation/runhook_test.go [hunk 1]
```diff
@@ -107,11 +107,11 @@ func (s *RunHookSuite) TestPrepareHookCtxError(c *gc.C) {
 }
 
 func (s *RunHookSuite) TestPrepareHookError_Run(c *gc.C) {
-	s.testPrepareHookError(c, (operation.Factory).NewRunHook, false, false)
+	s.testPrepareHookError(c, operation.Factory.NewRunHook, false, false)
```
Comment (manadart) [inline worker/uniter/operation/runhook_test.go]: All these one-liners are linter appeasement. Sorry for the noise.

**Comment by manadart** [comment]
@howbazaar I added a check to `beforeHook`. This appears to achieve what you describe.
The check prior to merge looks OK to keep; can you confirm?

File: worker/uniter/operation/runhook.go [hunk 2]
```diff
@@ -153,7 +153,16 @@ func (rh *runHook) beforeHook(state State) error {
 			Status: string(status.Maintenance),
 			Info:   "cleaning up prior to charm deletion",
 		})
+	case hooks.LeaderElected:
+		// Check if leadership has changed between queueing of the hook and
+		// Actual execution. Set the error if we are no longer the leader.
```
Comment (wallyworld) [inline worker/uniter/operation/runhook.go]: s/Actual/actual

File: worker/uniter/operation/runhook.go [hunk 3]
```diff
@@ -153,7 +153,16 @@ func (rh *runHook) beforeHook(state State) error {
...
+		isLeader := false
+		isLeader, err = rh.runner.Context().IsLeader()
+		if err == nil && !isLeader {
+			err = context.ErrIsNotLeader
```
Comment (howbazaar) [inline worker/uniter/operation/runhook.go]: I think that if the unit is no longer the leader, we don't want to leave the uniter in an error state.

Perhaps just log an info message saying skipping due leader-elected to no longer leader, and return nil?

Comment (manadart) [reply worker/uniter/operation/runhook.go]: @howbazaar, @wallyworld 

Can we return directly with a nil state, or should we behave as if the hook had run despite not calling `RunHook`? I.e do we:
- call `callbacks.NotifyHookCompleted`?
- call `afterHook`?
- return a changed state?

Apologies for the churn; first time in the uniter.

Comment (wallyworld) [reply worker/uniter/operation/runhook.go]: @manadart 
The correct approach is to modify the runhook Prepare() method to return ErrSkipExecute if the unit is no longer a leader. So take the logic you have here and move to Prepare(). That should be all you need to do

Comment (manadart) [reply worker/uniter/operation/runhook.go]: @wallyworld Suggested change is applied.

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**